karma/cogs/stats.py

```diff
@@ -0,0 +1,632 @@
+import re
+import io
+
+import logging
+import discord
+
+import matplotlib
+import matplotlib.pyplot as plt
+params = {
+    "text.color" : "w",
+    "ytick.color" : "w",
+    "xtick.color" : "w",
+    "axes.labelcolor" : "w",
+    "axes.edgecolor" : "w"
+}
+plt.rcParams.update(params)
+
+from discord.ext import commands
+
+from itertools import accumulate
+
+import karma.database as db
+
+logger = logging.getLogger(__name__)
+
+valid_types = ['all', 'given', 'received', 'bff', 'graph']
+
+class Stats(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.command(pass_context=True, help='Get the BFF score for two users (Person 1 -> Person 2).')
+    async def bff(self, ctx, uid1, uid2):
+
+        await ctx.channel.trigger_typing()
+        logger.info("Command invoked: bff")
+
+        guild = ctx.guild
+        pattern = re.compile(r'<@!?(?P<user_id>\d+)>')
+        match = pattern.search(uid1)
+        if match and guild.get_member(int(match.group('user_id'))):
+            match_other = pattern.search(uid2)
+            if match_other and guild.get_member(int(match_other.group('user_id'))):
+                bff, bff_score, err = calculate_bff(match.group('user_id'), uid_to_calculate=match_other.group('user_id'))
+                if err == None:
+                     await ctx.send("The BFF score for **{}** with **{}** is **{:.2f}**. Wow! :sparkling_heart:".format(
+                         get_username(match.group('user_id'), guild),
+                         get_username(match_other.group('user_id'), guild),
+                         bff_score
+                     ))
+                else:
+                    await ctx.send(err)
+            else:
+                await ctx.send('Could not find user: {}'.format(key))
+        else:
+            await ctx.send('Could not find user: {}'.format(key))
+
+    @commands.command(pass_context=True, help='Get stats for the server, or mention a user for specifics.\nValid types are {}'.format(valid_types))
+    async def stats(self, ctx, type, *args):
```

Comment:
I'd be careful using type as a variable name, I think it's one that might be reserved as a keyword, given its syntax highlighting

Reply:
Resolved desu